summary/summary.go

```diff
@@ -0,0 +1,147 @@
+	// latest derived minute, hour, and day stats. Updated every minute.
+	derivedStats info.DerivedStats // Guarded by dataLock.
+	dataLock     sync.RWMutex
+}
+
+// Adds a new seconds sample.
+// If enough seconds samples are collected, a minute sample is generated and derived
+// stats are updated.
+func (s *StatsSummary) AddSample(stat info.ContainerStats) error {
+	sample := secondSample{}
+	if s.available.Cpu {
+		sample.Cpu = stat.Cpu.Usage.Total
+	}
+	if s.available.Memory {
+		sample.Memory = stat.Memory.WorkingSet
+	}
+	s.secondSamples = append(s.secondSamples, &sample)
+	// TODO(jnagal): Use 'available' to avoid unnecessary computation.
+	if len(s.secondSamples) == 60 {
+		// Make a minute stat.
+		minuteSample := GetMinutePercentiles(s.secondSamples)
+		// clear seconds samples.
+		s.secondSamples = s.secondSamples[:0]
+		s.minuteSamples.Add(minuteSample)
+		err := s.updateDerivedStats()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Generate new derived stats based on current minute stats samples.
+func (s *StatsSummary) updateDerivedStats() error {
+	derived := info.DerivedStats{}
+	derived.Timestamp = time.Now()
+	minuteSamples := s.minuteSamples.RecentStats(1)
+	if len(minuteSamples) != 1 {
+		return fmt.Errorf("failed to retrieve minute stats")
+	}
+	derived.MinuteUsage = *minuteSamples[0]
+	hourUsage, err := s.getDerivedUsage(60)
+	if err != nil {
+		return fmt.Errorf("failed to compute hour stats: %v", err)
+	}
+	dayUsage, err := s.getDerivedUsage(60 * 24)
+	if err != nil {
+		return fmt.Errorf("failed to compute day usage: %v", err)
+	}
+	derived.HourUsage = hourUsage
+	derived.DayUsage = dayUsage
+
+	s.dataLock.Lock()
+	defer s.dataLock.Unlock()
+	s.derivedStats = derived
+
+	return nil
+}
+
+// helper method to get hour and daily derived stats
+func (s *StatsSummary) getDerivedUsage(n int) (info.Usage, error) {
+	if n < 1 {
+		return info.Usage{}, fmt.Errorf("invalid number of samples requested: %d", n)
+	}
+	samples := s.minuteSamples.RecentStats(n)
+	numSamples := len(samples)
+	if numSamples < 1 {
+		return info.Usage{}, fmt.Errorf("failed to retrieve any minute stats.")
+	}
+	// We generate derived stats even with partial data.
+	usage := GetDerivedPercentiles(samples)
+	// Assumes we have equally placed minute samples.
+	usage.PercentComplete = int32(numSamples / n)
+	return usage, nil
+}
+
+// Return the latest calculated derived stats.
+func (s *StatsSummary) DerivedStats() (info.DerivedStats, error) {
+	s.dataLock.RLock()
+	defer s.dataLock.RUnlock()
+
+	return s.derivedStats, nil
+}
+
+func New(spec info.ContainerSpec) (*StatsSummary, error) {
+	summary := StatsSummary{}
+	if spec.HasCpu {
+		summary.available.Cpu = true
+	}
+	if spec.HasMemory {
+		summary.available.Memory = true
+	}
+	if !summary.available.Cpu && !summary.available.Memory {
+		return nil, fmt.Errorf("none of the resources are being tracked.")
+	}
+	summary.minuteSamples = NewSamplesBuffer(24 * 60 /* one day */)
```

Comment:
We spoke of doing the decay from seconds -> minutes -> hours -> day. How come we decided to just keep all the minutes instead?